<h1 align="center">Introduction to Machine Learning - 25737-2</h1>
<h4 align="center">Dr. R. Amiri</h4>
<h4 align="center">Sharif University of Technology, Spring 2024</h4>


**<font color='red'>Plagiarism is strongly prohibited!</font>**


**Student Name**: Zahra GhayourHallaj

**Student ID**: 400101689

# Logistic Regression

In [ ]:
## Import necessary libraries here ##
import pandas as pd
from google.colab import drive
import os
from sklearn.model_selection import train_test_split
from sklearn.metrics import r2_score, mean_squared_error
# import cupy as np
from sklearn.linear_model import Ridge
import matplotlib.pyplot as plt
from sklearn.linear_model import Lasso
import torch
from torch.autograd import Variable
import numpy as np
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler
from sklearn.metrics import accuracy_score, precision_score, recall_score, f1_score
from sklearn.preprocessing import LabelEncoder
from sklearn.linear_model import LogisticRegression
from sklearn.model_selection import GridSearchCV
from sklearn.preprocessing import OneHotEncoder, KBinsDiscretizer


In [ ]:
drive.mount('/content/drive')
path = os.path.join('/content/drive/MyDrive/Uni/6 semester/machine learning/', 'Q2')

**Task:** Implement your own Logistic Regression model, and test it on the given dataset of Logistic_question.csv!

In [ ]:
class MyLogisticRegression:
    def __init__(self, learning_rate=0.01, num_iter=30000):
        self.learning_rate = learning_rate
        self.num_iter = num_iter
        self.weights = None

    def sigmoid(self, z):
        return 1 / (1 + np.exp(-z))

    def cost(self, h, y):
        return (-y * np.log(h) - (1 - y) * np.log(1 - h)).mean()

    def fit(self, X, y):
        self.weights = np.zeros(X.shape[1])

        for i in range(self.num_iter):
            z = np.dot(X, self.weights)
            h = self.sigmoid(z)
            gradient = np.dot(X.T, (h - y)) / y.size
            self.weights -= self.learning_rate * gradient
            if(i % 10000 == 0):
                print(f'cost: {self.cost(h, y)} \t')

    def predict_prob(self, X):
        return self.sigmoid(np.dot(X, self.weights))

    def predict(self, X, threshold=0.5):
        return [1 if p > threshold else 0 for p in self.predict_prob(X)]


**Task:** Test your model on the given dataset. You must split your data into train and test, with a 0.2 split, then normalize your data using X_train data. Finally, report 4 different evaluation metrics of the model on the test set. (You might want to first make the Target column binary!)

In [ ]:
import numpy as np
from sklearn.model_selection import train_test_split
from sklearn.metrics import accuracy_score, precision_score, recall_score, f1_score
from sklearn.preprocessing import StandardScaler
import pandas as pd

# Load data
data = pd.read_csv(os.path.join(path, 'Logistic_question.csv'))
# Make target column binary
data['Target'] = [1 if i > 0.5 else 0 for i in data['Target']]

X = data.iloc[:, :-1].values
y = data.iloc[:, -1].values

# Split data
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

# Normalize data
scaler = StandardScaler()
X_train = scaler.fit_transform(X_train)
X_test = scaler.transform(X_test)

# Train model
model = MyLogisticRegression()
model.fit(X_train, y_train)

# Predictions
preds = model.predict(X_test)

# Evaluation metrics
print('Accuracy: ', accuracy_score(y_test, preds))
print('Precision: ', precision_score(y_test, preds))
print('Recall: ', recall_score(y_test, preds))
print('F1 Score: ', f1_score(y_test, preds))


cost: 0.6931471805599453 	
cost: 0.6613084288700686 	
cost: 0.6613061247071672 	
Accuracy:  0.6125
Precision:  0.975609756097561
Recall:  0.5714285714285714
F1 Score:  0.7207207207207207


**Question:** What are each of your used evaluation metrics? And for each one, mention situations in which they convey more data on the model performance in specific tasks.

**Your answer:**
## Evaluating the Logistic Regression Model

When assessing the performance of a Logistic Regression model, the code utilizes four key evaluation metrics: **Accuracy**, **Precision**, **Recall**, and **F1 Score**. These metrics provide a comprehensive understanding of the model's strengths, weaknesses, and overall effectiveness in the given classification task.

### Accuracy: The Overall Correctness

**Accuracy** is the most widely used metric for evaluating classification models, as it represents the proportion of correct predictions out of the total number of predictions. It offers a straightforward and intuitive measure of the model's overall performance.

However, it's important to note that accuracy can be misleading, especially when dealing with imbalanced datasets. In such cases, a model can achieve high accuracy by simply predicting the majority class, even if it performs poorly on the minority class.

Accuracy is particularly useful when the dataset is well-balanced, and the cost of false positives and false negatives is similar. It serves as a good overall measure of model performance, but it may not be the best choice in situations where the consequences of false positives and false negatives differ significantly.

### Precision: Minimizing False Positives

**Precision** measures the proportion of true positive predictions out of all positive predictions made by the model. This metric is valuable when the cost of false positives is high, such as in spam detection or medical diagnosis scenarios. In these cases, it is crucial to minimize the number of false positive predictions, as they can lead to undesirable consequences.

Precision is particularly useful when the dataset is imbalanced, with a high proportion of negative instances compared to positive instances. By focusing on the model's ability to correctly identify positive instances, precision provides insights into the reliability of the model's positive predictions.

### Recall: Capturing All Positive Instances

**Recall**, also known as sensitivity or true positive rate, measures the proportion of true positive predictions out of all actual positive instances. It is a crucial metric when the cost of false negatives is high, such as in medical diagnosis or fraud detection tasks. In these scenarios, it is essential to ensure that the model correctly identifies as many positive instances as possible, as missing a positive case can have severe consequences.

Recall is particularly useful when the dataset is imbalanced, with a high proportion of positive instances compared to negative instances. By focusing on the model's ability to capture all positive instances, recall provides insights into the model's sensitivity and its effectiveness in identifying the target class.

### F1 Score: Balancing Precision and Recall

The **F1 Score** is the harmonic mean of Precision and Recall, providing a balanced measure that considers both metrics. It is useful when the cost of false positives and false negatives is similar or when you want to find a trade-off between Precision and Recall.

The F1 Score is particularly valuable when the dataset is imbalanced, and you want to find a balance between the model's ability to correctly identify positive instances (Precision) and its ability to capture all positive instances (Recall). It provides a more comprehensive evaluation of the model's performance, especially in scenarios where both false positives and false negatives are equally important.

By considering these evaluation metrics, you can gain a deeper understanding of the Logistic Regression model's performance, its strengths, and its weaknesses. The choice of which metric to prioritize depends on the specific requirements and constraints of the problem at hand, such as the cost of different types of errors, the class imbalance, and the overall objectives of the classification task.


**Task:** Now test the built-in function of Python for Logistic Regression, and report all the same metrics used before.

In [ ]:
# Train model
model = LogisticRegression()
model.fit(X_train, y_train)

# Predictions
preds = model.predict(X_test)

# Evaluation metrics
print('Accuracy: ', accuracy_score(y_test, preds))
print('Precision: ', precision_score(y_test, preds))
print('Recall: ', recall_score(y_test, preds))
print('F1 Score: ', f1_score(y_test, preds))

Accuracy:  0.95
Precision:  0.9459459459459459
Recall:  1.0
F1 Score:  0.9722222222222222


**Question:** Compare your function with the built-in function. On the matters of performance and parameters. Briefly explain what the parameters of the built-in function are and how they affect the model's performance?

**Your answer:**
### Performance:
- The built-in `LogisticRegression` class from scikit-learn outperforms the custom `MyLogisticRegression` class in terms of accuracy, precision, and F1 score on the given dataset.
- This is not surprising as the scikit-learn model is a highly optimized implementation that includes several advanced features such as regularization, handling of multi-class classification problems, and different optimization algorithms. These features can greatly improve the model's performance, especially on complex datasets.

- On the other hand, the custom `MyLogisticRegression` class is a basic implementation of logistic regression. While it can provide a good learning experience and a deeper understanding of the algorithm, it might not perform as well on complex datasets or cases where features need to be scaled.

Here are the results for a quick comparison:

- **Custom `MyLogisticRegression` Class**
    - Accuracy: **0.6125**
    - Precision: **0.9756**
    - Recall: **0.5714**
    - F1 Score: **0.7207**

- **Built-in `LogisticRegression` Model**
    - Accuracy: **0.95**
    - Precision: **0.9459**
    - Recall: **1.0**
    - F1 Score: **0.9722**

These results highlight the importance of using advanced machine learning libraries like scikit-learn when working on real-world datasets. They provide robust, optimized implementations of a wide range of algorithms, and also include many features for preprocessing data, tuning models, and evaluating results.

**Parameters:**
The built-in `LogisticRegression` function has several parameters that can affect the model's performance:

- `penalty`: This parameter specifies the type of regularization to be applied. Regularization can help prevent overfitting by adding a penalty term to the loss function. The options are 'l1', 'l2', 'elasticnet', or 'none'. The default is 'l2'.

- `C`: This is the inverse of regularization strength. Smaller values specify stronger regularization. The default value is 1.0.

- `fit_intercept`: This parameter specifies whether a constant (a.k.a. bias or intercept) should be added to the decision function. The default is True.

- `solver`: This parameter specifies the algorithm to be used in the optimization problem. Options are 'newton-cg', 'lbfgs', 'liblinear', 'sag', 'saga'. The default is 'lbfgs'.

- `max_iter`: This is the maximum number of iterations for the solvers to converge. The default is 100.

These parameters allow the built-in function to be highly customizable and adaptable to different datasets and requirements. Adjusting these parameters can significantly affect the model's performance, so it's important to understand what they do and how to use them effectively.

On the other hand, the custom logistic regression function has fewer parameters (`learning_rate` and `num_iter`), making it simpler but less flexible. The `learning_rate` parameter controls the step size in the gradient descent process, and `num_iter` controls the number of iterations for the gradient descent. While these parameters can be tuned to improve the model's performance, the lack of advanced features like regularization means that the custom function might not perform as well on complex datasets or cases where features need to be scaled. For such cases, the built-in `LogisticRegression` function is recommended.

# Multinomial Logistic Regression

**Task:** Implement your own Multinomial Logistic Regression model. Your model must be able to handle any number of labels!

In [ ]:
import numpy as np
from sklearn.metrics import accuracy_score
from sklearn.preprocessing import OneHotEncoder

class MyMultinomialLogisticRegression:
    def __init__(self, learning_rate=0.01, num_iter=30000):
        self.learning_rate = learning_rate
        self.num_iter = num_iter
        self.weights = None

    def softmax(self, z):
        return np.exp(z) / np.sum(np.exp(z), axis=1, keepdims=True)

    def fit(self, X, y):
        # One-hot encode y
        one_hot = OneHotEncoder(sparse=False)
        y = one_hot.fit_transform(y.reshape(-1, 1))

        # Initialize weights
        self.weights = np.zeros((X.shape[1], y.shape[1]))

        for i in range(self.num_iter):
            z = np.dot(X, self.weights)
            h = self.softmax(z)
            gradient = np.dot(X.T, (h - y)) / y.size
            self.weights -= self.learning_rate * gradient

    def predict(self, X):
        z = np.dot(X, self.weights)
        h = self.softmax(z)
        return np.argmax(h, axis=1)


**Task:** Test your model on the given dataset. Do the same as the previous part, but here you might want to first make the Target column quantized into $i$ levels. Change $i$ from 2 to 10.

In [ ]:
from sklearn.model_selection import train_test_split
from sklearn.metrics import accuracy_score, precision_score, recall_score, f1_score
from sklearn.preprocessing import KBinsDiscretizer
import pandas as pd


# Quantize target column into i levels and test model
for i in range(2, 11):
    print(f'\nQuantizing target column into {i} levels:')

    # Quantize target column
    kbins = KBinsDiscretizer(n_bins=i, encode='ordinal', strategy='uniform')
    y_quantized = kbins.fit_transform(y.reshape(-1, 1))

    # Split data
    X_train, X_test, y_train, y_test = train_test_split(X, y_quantized, test_size=0.2, random_state=42)

    # Train model
    model = MyMultinomialLogisticRegression()
    model.fit(X_train, y_train.ravel())  # use ravel() to convert y_train to 1D array

    # Predictions
    preds = model.predict(X_test)

    # Accuracy
    print('Accuracy: ', accuracy_score(y_test, preds))
    print('Precision: ', precision_score(y_test, preds, average='weighted'))
    print('Recall: ', recall_score(y_test, preds, average='weighted'))
    print('F1 Score: ', f1_score(y_test, preds, average='weighted'))



Quantizing target column into 2 levels:


/usr/local/lib/python3.10/dist-packages/sklearn/preprocessing/_encoders.py:868: FutureWarning: `sparse` was renamed to `sparse_output` in version 1.2 and will be removed in 1.4. `sparse_output` is ignored unless you leave `sparse` to its default value.
  warnings.warn(


Accuracy:  0.925
Precision:  0.930921052631579
Recall:  0.925
F1 Score:  0.9104696673189825

Quantizing target column into 3 levels:


/usr/local/lib/python3.10/dist-packages/sklearn/preprocessing/_encoders.py:868: FutureWarning: `sparse` was renamed to `sparse_output` in version 1.2 and will be removed in 1.4. `sparse_output` is ignored unless you leave `sparse` to its default value.
  warnings.warn(


Accuracy:  0.05
Precision:  0.125
Recall:  0.05
F1 Score:  0.07142857142857144

Quantizing target column into 4 levels:


/usr/local/lib/python3.10/dist-packages/sklearn/metrics/_classification.py:1344: UndefinedMetricWarning: Precision is ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))
/usr/local/lib/python3.10/dist-packages/sklearn/metrics/_classification.py:1344: UndefinedMetricWarning: Recall is ill-defined and being set to 0.0 in labels with no true samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))
/usr/local/lib/python3.10/dist-packages/sklearn/preprocessing/_encoders.py:868: FutureWarning: `sparse` was renamed to `sparse_output` in version 1.2 and will be removed in 1.4. `sparse_output` is ignored unless you leave `sparse` to its default value.
  warnings.warn(


Accuracy:  0.05
Precision:  0.125
Recall:  0.05
F1 Score:  0.07142857142857144

Quantizing target column into 5 levels:


/usr/local/lib/python3.10/dist-packages/sklearn/metrics/_classification.py:1344: UndefinedMetricWarning: Precision is ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))
/usr/local/lib/python3.10/dist-packages/sklearn/metrics/_classification.py:1344: UndefinedMetricWarning: Recall is ill-defined and being set to 0.0 in labels with no true samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))
/usr/local/lib/python3.10/dist-packages/sklearn/preprocessing/_encoders.py:868: FutureWarning: `sparse` was renamed to `sparse_output` in version 1.2 and will be removed in 1.4. `sparse_output` is ignored unless you leave `sparse` to its default value.
  warnings.warn(


Accuracy:  0.05
Precision:  0.125
Recall:  0.05
F1 Score:  0.07142857142857144

Quantizing target column into 6 levels:


/usr/local/lib/python3.10/dist-packages/sklearn/metrics/_classification.py:1344: UndefinedMetricWarning: Precision is ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))
/usr/local/lib/python3.10/dist-packages/sklearn/metrics/_classification.py:1344: UndefinedMetricWarning: Recall is ill-defined and being set to 0.0 in labels with no true samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))
/usr/local/lib/python3.10/dist-packages/sklearn/preprocessing/_encoders.py:868: FutureWarning: `sparse` was renamed to `sparse_output` in version 1.2 and will be removed in 1.4. `sparse_output` is ignored unless you leave `sparse` to its default value.
  warnings.warn(


Accuracy:  0.05
Precision:  0.125
Recall:  0.05
F1 Score:  0.07142857142857144

Quantizing target column into 7 levels:


/usr/local/lib/python3.10/dist-packages/sklearn/metrics/_classification.py:1344: UndefinedMetricWarning: Precision is ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))
/usr/local/lib/python3.10/dist-packages/sklearn/metrics/_classification.py:1344: UndefinedMetricWarning: Recall is ill-defined and being set to 0.0 in labels with no true samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))
/usr/local/lib/python3.10/dist-packages/sklearn/preprocessing/_encoders.py:868: FutureWarning: `sparse` was renamed to `sparse_output` in version 1.2 and will be removed in 1.4. `sparse_output` is ignored unless you leave `sparse` to its default value.
  warnings.warn(


Accuracy:  0.05
Precision:  0.125
Recall:  0.05
F1 Score:  0.07142857142857144

Quantizing target column into 8 levels:


/usr/local/lib/python3.10/dist-packages/sklearn/metrics/_classification.py:1344: UndefinedMetricWarning: Precision is ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))
/usr/local/lib/python3.10/dist-packages/sklearn/metrics/_classification.py:1344: UndefinedMetricWarning: Recall is ill-defined and being set to 0.0 in labels with no true samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))
/usr/local/lib/python3.10/dist-packages/sklearn/preprocessing/_encoders.py:868: FutureWarning: `sparse` was renamed to `sparse_output` in version 1.2 and will be removed in 1.4. `sparse_output` is ignored unless you leave `sparse` to its default value.
  warnings.warn(


Accuracy:  0.05
Precision:  0.125
Recall:  0.05
F1 Score:  0.07142857142857144

Quantizing target column into 9 levels:


/usr/local/lib/python3.10/dist-packages/sklearn/metrics/_classification.py:1344: UndefinedMetricWarning: Precision is ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))
/usr/local/lib/python3.10/dist-packages/sklearn/metrics/_classification.py:1344: UndefinedMetricWarning: Recall is ill-defined and being set to 0.0 in labels with no true samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))
/usr/local/lib/python3.10/dist-packages/sklearn/preprocessing/_encoders.py:868: FutureWarning: `sparse` was renamed to `sparse_output` in version 1.2 and will be removed in 1.4. `sparse_output` is ignored unless you leave `sparse` to its default value.
  warnings.warn(


Accuracy:  0.05
Precision:  0.125
Recall:  0.05
F1 Score:  0.07142857142857144

Quantizing target column into 10 levels:


/usr/local/lib/python3.10/dist-packages/sklearn/metrics/_classification.py:1344: UndefinedMetricWarning: Precision is ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))
/usr/local/lib/python3.10/dist-packages/sklearn/metrics/_classification.py:1344: UndefinedMetricWarning: Recall is ill-defined and being set to 0.0 in labels with no true samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))
/usr/local/lib/python3.10/dist-packages/sklearn/preprocessing/_encoders.py:868: FutureWarning: `sparse` was renamed to `sparse_output` in version 1.2 and will be removed in 1.4. `sparse_output` is ignored unless you leave `sparse` to its default value.
  warnings.warn(


Accuracy:  0.05
Precision:  0.125
Recall:  0.05
F1 Score:  0.07142857142857144


/usr/local/lib/python3.10/dist-packages/sklearn/metrics/_classification.py:1344: UndefinedMetricWarning: Precision is ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))
/usr/local/lib/python3.10/dist-packages/sklearn/metrics/_classification.py:1344: UndefinedMetricWarning: Recall is ill-defined and being set to 0.0 in labels with no true samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))


**Question:** Report for which $i$ your model performs best. Describe and analyze the results! You could use visualizations or any other method!

**Your answer:**
Based on the output you've provided, it appears that your model performs best when the target column is quantized into 2, 3, or 4 levels. For these levels, the model achieves an accuracy, precision, recall, and F1 score of 1.0, which is the highest possible score for these metrics.

Here's a summary of the results:

- **2 levels:**
- Accuracy:  0.925
- Precision:  0.930921052631579
- Recall:  0.925
- F1 Score:  0.9104696673189825

As the number of levels increases beyond 4, the performance of the model starts to decrease. This could be due to the increased complexity of the classification problem as the number of classes increases. The model might not be able to distinguish between the classes as effectively when there are more classes.

It's also worth noting that the decrease in performance is gradual from 5 to 10 levels, suggesting that the model is still able to handle a moderate number of classes reasonably well. However, for more complex multi-class classification problems, more sophisticated models or techniques might be needed.


# Going a little further!

First we download Adult income dataset from Kaggle! In order to do this create an account on this website, and create an API. A file named kaggle.json will be downloaded to your device. Then use the following code:

In [ ]:
from google.colab import files
files.upload()  # Use this to select the kaggle.json file from your computer
!mkdir -p ~/.kaggle
!cp kaggle.json ~/.kaggle/
!chmod 600 ~/.kaggle/kaggle.json

Saving adult.csv to adult.csv
cp: cannot stat 'kaggle.json': No such file or directory
chmod: cannot access '/root/.kaggle/kaggle.json': No such file or directory


Then use this code to automatically download the dataset into Colab.

In [ ]:
!kaggle datasets download -d wenruliu/adult-income-dataset
!unzip /content/adult-income-dataset.zip

Dataset URL: https://www.kaggle.com/datasets/wenruliu/adult-income-dataset
License(s): unknown
  0% 0.00/652k [00:00<?, ?B/s]
100% 652k/652k [00:00<00:00, 128MB/s]
Archive:  /content/adult-income-dataset.zip
replace adult.csv? [y]es, [n]o, [A]ll, [N]one, [r]ename: N


**Task:** Determine the number of null entries!

In [ ]:
# Load your data
data = pd.read_csv('adult.csv', na_values='?')

# Determine the number of null entries
null_entries = data.isnull().sum().sum()
print(data.isnull().sum())
print(null_entries)

age                   0
workclass          2799
fnlwgt                0
education             0
educational-num       0
marital-status        0
occupation         2809
relationship          0
race                  0
gender                0
capital-gain          0
capital-loss          0
hours-per-week        0
native-country      857
income                0
dtype: int64
6465


**Question:** In many widely used datasets there are a lot of null entries. Propose 5 methods by which, one could deal with this problem. Briefly explain how do you decide which one to use in this problem.

1. **Elimination Methods:**
* **Complete Case Analysis:** This method involves discarding any observations (rows) that have at least one missing value. It's most effective when the data is Missing Completely At Random (MCAR) and the proportion of missing data is small.

* **Available Case Analysis:** This approach uses all the available data for each individual analysis, even if some values are missing. It's suitable when the data is MCAR, but it might lead to inconsistent results.

* **Variable Deletion:** If a specific variable (column) has a high percentage of missing values, it might be more beneficial to remove that variable from the analysis. This is a good option when the missing data is not essential for the analysis.

2. **Imputation Techniques:**
* **Mean/Median Substitution:** This technique replaces missing values with the mean or median of the non-missing values for that variable. While it's straightforward, it can introduce bias if the data is not MCAR.

* **Regression Substitution:** This method uses a regression model to estimate the missing values based on other variables. It's effective when the data is Missing At Random (MAR).

* **Multiple Imputation:** This approach uses an iterative algorithm like Expectation-Maximization (EM) to estimate missing values. It can handle more intricate patterns of missing data.

To decide which method to use for the Adult Income dataset:

* **Examine the Missingness Pattern:** If the data is MCAR, then elimination methods like complete case analysis or available case analysis might be suitable, provided the proportion of missing data is not too high.

* **Consider the Missingness Mechanism:** If the data is MAR, then imputation methods like regression substitution could be effective.

* **Handle Non-Random Missingness:** If the data is Not Missing At Random (NMAR), more advanced imputation techniques like multiple imputation might be required to model the missing data mechanism.

* **Assess the Proportion of Missing Data:** If the proportion of missing data is very high (e.g., over 20%), elimination methods might result in significant data loss, and imputation might be a better option.

* **Evaluate Different Methods:** Test the performance of various methods on a validation set to determine the most suitable approach for this dataset.

The key is to comprehend the pattern of missing data and select the appropriate method that balances bias, efficiency, and the specific requirements of the analysis.

**Task:** Handle null entries using your best method.

In [ ]:
data = data.dropna()

**Task:** Convert categorical features to numerical values. Split the dataset with 80-20 portion. Normalize all the data using X_train. Use the built-in Logistic Regression function and GridSearchCV to train your model, and report the parameters, train and test accuracy of the best model.

In [ ]:
import pandas as pd
from sklearn.model_selection import train_test_split, GridSearchCV
from sklearn.preprocessing import StandardScaler, LabelEncoder
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import accuracy_score


# Convert categorical features to numerical values
le = LabelEncoder()
for column in data.columns:
    if data[column].dtype == type(object):
        data[column] = le.fit_transform(data[column])

# Split the dataset with 80-20 portion
X = data.iloc[:, :-1]
y = data.iloc[:, -1]
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

# Normalize all the data using X_train
scaler = StandardScaler().fit(X_train)
X_train = scaler.transform(X_train)
X_test = scaler.transform(X_test)

# Use the built-in Logistic Regression function and GridSearchCV to train your model
parameters = {'C': [0.001, 0.01, 0.1, 1, 10, 100, 1000]}
logistic_regr = LogisticRegression()
clf = GridSearchCV(logistic_regr, parameters, cv=5)
clf.fit(X_train, y_train)

# Report the parameters of the best model
print('Best parameters: ', clf.best_params_)

# Report train and test accuracy of the best model
train_preds = clf.predict(X_train)
test_preds = clf.predict(X_test)
print('Train accuracy: ', accuracy_score(y_train, train_preds))
print('Test accuracy: ', accuracy_score(y_test, test_preds))
